### PR TITLE
High-speed gun inaccuracy modifiers

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

- Implements weapon inaccuracy based on movement speeds.
- Travelling at normal run speed will result in no additional accuracy penalty.
- Travelling at speeds above default mob running speed will decrease your accuracy by 10 degrees per movement level (Captain's jetpack speed is 1 movement speed increase, so that speed is 10 degrees penalty).
  - The above penalty is doubled if you are in space and don't have a lattice or wall to hold onto.
  - The above penalty is not applied if you haven't moved in the past 1 second.
- Travelling at speed lower than default run speed will increase your accuracy, giving you 8 degrees of bonus for every point slower you are than runspeed.
  - This means by walking you will gain 12 points of accuracy.
  - This applies even if you are not moving.
  - This only affects the poor aimer trait and holding a weapon with 1 hand. This will reduce the spread of energy guns from 30 down to 18 if you are walking.
- Fixes unwielded weapon spread being randomised too early, resulting in double randomisation.
- Buffs the unwielded inaccuracy of all guns to account for the increased spread randomisation from the fix.

**Code notes:**
- This makes last_move_time actually have uses. Previously it was defined but never assigned to.

## Why It's Good For The Game

Ranged space combat involving jetpacks often ends up looking very unfair when you see someone holding onto the station against someone moving at high speeds in space. When 1 party has a jetpack, they get free mobility over speed and increased movement speed over their opponent making it completely impossible to escape someone with a jetpack, but also extremely difficult to fight back against them. Often when fighting in space without a jetpack against someone who has one, you have no way to find cover but your opponent moves so fast they are very difficult to hit.
This PR makes it so that jetpacking around freely in space will make you super inaccurate unless you cuddle up against the wall, enable stabilisation and stand still or use a slow moving jetpack.
Jetpack gun combat was the main thing I wanted to make less effective in this PR.

This also makes it harder to use weapons when you have extremely high movement speeds, since fighting against a ranged opponent who can move faster than bullets is not only unfun but feels very unfair.

This PR also gives benefits to playing slowly in gunfights, allowing you to have more accuracy when walking which will give you a bonus if you are holding a shield and playing more slowly which is good since slower paced gunfights and combat often leads to more interesting gameplay as the fight isn't over in as quick time and the other party has a chance to get away.

**This will not affect weapon accuracy of people running at run speed. This only applies a penalty if you have movement bonuses that make you faster than normal runspeed.**

## Testing Photographs and Procedure

https://github.com/BeeStation/BeeStation-Hornet/assets/26465327/2c724ad8-ada5-4852-89ba-1aa6ad155e2e

(Should have shown my hands)
https://github.com/BeeStation/BeeStation-Hornet/assets/26465327/9cf4e2fd-f2f9-4327-a94f-539fad5a422b

## Changelog
:cl:
balance: Guns now have reduced accuracy when moving faster than the normal runspeed, and when moving at high speeds with nothing to hold onto in space. Walking will now make dual-wielding and holding a gun and shield slightly more accurate.
code: Updates the comments attached to buckling to be correct
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Note that rand(rand(n)) will result in 1 having a n probability of occuring while n has a probability of 1 occuring. The probability of m occuring is m * (n - m). This means that values closer to 0 are way more likely than values in the fringe. I have made the new code with the random fix halved, although in reality the average will be 40% but at the same time the chance for the weapon to be wildly inaccurate is completely removed so it should be pretty similar to previous.

File: code/modules/mob/living/living_defense.dm [hunk 1]
```diff
@@ -437,3 +437,40 @@
...
+		. += weapon.spread_unwielded
+	// Nothing to hold onto, slight penalty for flying around in space
+	var/default_speed = get_config_multiplicative_speed() + CONFIG_GET(number/movedelay/run_delay)
+	var/current_speed = cached_multiplicative_slowdown || total_multiplicative_slowdown()
+	// Lower speed is better, so this is reversed
+	var/speed_delta = default_speed - current_speed
+	// Are we holding onto something?
+	var/is_holding = has_gravity(get_turf(src))
+	if (!is_holding)
+		if(locate(/obj/structure/lattice) in range(1, get_turf(src))) //Not realistic but makes pushing things in space easier
```
Comment (PowerfulBacon) [inline code/modules/mob/living/living_defense.dm]:
```suggestion
		if(locate(/obj/structure/lattice) in range(1, get_turf(src)))
```

File: code/modules/mob/living/living_defense.dm [hunk 2]
```diff
@@ -437,3 +437,40 @@
...
+
+/mob/living/proc/get_weapon_inaccuracy_modifier(atom/target, obj/item/gun/weapon)
+	. = 0
+	if(HAS_TRAIT(src, TRAIT_POOR_AIM)) //nice shootin' tex
+		. += 25
+	// Unwielded weapons
+	if(!weapon.is_wielded && weapon.requires_wielding)
+		. += weapon.spread_unwielded
+	// Nothing to hold onto, slight penalty for flying around in space
+	var/default_speed = get_config_multiplicative_speed() + CONFIG_GET(number/movedelay/run_delay)
```
Comment (PowerfulBacon) [inline code/modules/mob/living/living_defense.dm:449]: get_config_multiplicative_speed() doesn't take into account the run delay, so we have to add it on here.

File: code/modules/mob/living/living_defense.dm [hunk 3]
```diff
@@ -437,3 +437,40 @@
...
+	if (speed_delta > 0)
+		// We haven't moved in 1 second, give us no penalty to aiming
+		if (last_move_time + 1 SECONDS < world.time)
+			return
+		. += (speed_delta * 10) * (is_holding ? 1 : 2)
+	else
+		// Can only improve up to the maximum improvement, otherwise shotguns gain accuracy when walking
+		// This means walking will improve your accuracy by a total of 12.
+		// This is only really useful if you are using a gun with a shield.
+		. = max(0, . + speed_delta * 8)
```
Comment (PowerfulBacon) [inline code/modules/mob/living/living_defense.dm:476]: This provides a bonus, since speed_delta is negative here

**Comment by PowerfulBacon** [comment]
https://github.com/BeeStation/BeeStation-Hornet/assets/26465327/bee50daf-90eb-4748-b463-81e7a249b6af

This now also applies to vehicles.

**Comment by PowerfulBacon** [comment]
Note that due to a bug fix this may result in the new pocket pistol having greater spread than it did before.

**Comment by DarnTheMarn** [comment]
Does this also apply to things where you are buckled in, like on a chair being drug by someone faster than walking speed or the secway? Thought about this since with armor slowdown there is 0 reason to not use the secway permanently now since it makes you harder to hit, prevents some slips, and has a flat move speed that isn't affected by armor

**Review by itsmeow — APPROVED** [review]

**Review by Rukofamicom — APPROVED** [review]

**Pull request merged**